### PR TITLE
JBTM-2363 Blacktie c++ tx client build build failure

### PR DESCRIPTION
**Pull request by zhfeng**
NO_TEST

### CODE REVIEW

File: blacktie/tx/src/test/cpp/TestTransactions.cxx [hunk 1]
```diff
@@ -527,11 +527,14 @@ void TestTransactions::test_wait_for_recovery()
...
-	while ((rcCnt2 == 0 || (nrecs = count_log_records()) == nrecs1) && nsecs != 0) {
-		if (nsecs-- % 10 == 0) {
-			btlogger("TestTransactions::test_run_recovery sleeping for %d seconds", nsecs);
+	nrecs = count_log_records();
+	if(nrecs > 0 ) {
+		while ((rcCnt2 == 0 || (nrecs = count_log_records()) == nrecs1) && nsecs != 0) {
+			if (nsecs-- % 10 == 0) {
+				btlogger("TestTransactions::test_run_recovery sleeping for %d seconds", nsecs);
+			}
+			doSix(1);
```
Comment (mmusgrov) [inline blacktie/tx/src/test/cpp/TestTransactions.cxx]: The previous test makes no sense unless you move the sleep (doSix) inside the if block

File: blacktie/tx/src/test/cpp/TestTransactions.cxx [hunk 2]
```diff
@@ -527,11 +527,14 @@ void TestTransactions::test_wait_for_recovery()
 	(void) dummy_rm_add_fault(fault2);
 	(void) dummy_rm_add_fault(fault3);
 	BT_ASSERT_EQUAL(TX_OK, tx_open());
-	while ((rcCnt2 == 0 || (nrecs = count_log_records()) == nrecs1) && nsecs != 0) {
-		if (nsecs-- % 10 == 0) {
-			btlogger("TestTransactions::test_run_recovery sleeping for %d seconds", nsecs);
+	nrecs = count_log_records();
+	if(nrecs > 0 ) {
+		while ((rcCnt2 == 0 || (nrecs = count_log_records()) == nrecs1) && nsecs != 0) {
```
Comment (mmusgrov) [inline blacktie/tx/src/test/cpp/TestTransactions.cxx]: This test is waiting for recovery to happen. The test for recovery is the negation of the first condition:
    (rcCnt2 == 0 || (nrecs = count_log_records()) == nrecs1)
I think the intent of the second condition (ie nsecs != 0) is to only wait for a maximum of nsecs seconds before giving up. But the way the if condition is written means that the loop will always wait for nsecs but the loop should terminate as soon as recovery is detected (ie when the first condition is false). Why not something similar to:

		const int SLEEP_INTERVAL = 10;
		
		for (int i = 0; i < nsecs; i += SLEEP_INTERVAL) {
			if (rcCnt2 == 0 || (nrecs = count_log_records()) == nrecs1)
			    break;
			btlogger("TestTransactions::test_run_recovery sleeping for another %d seconds", SLEEP_INTERVAL);
			doSix(SLEEP_INTERVAL)
		}

File: blacktie/tx/src/main/cpp/HttpTxManager.cxx [hunk 3]
```diff
@@ -207,7 +207,7 @@ bool HttpTxManager::recover(XAWrapper *resource)
 	if (ri.status_code == 200) {
 		LOG4CXX_DEBUG(httptxlogger, "recovery: told TM about participant "
 			<< resource->get_name() << " xid: " << xid);
-
+		_branches.erase(resource->get_name());
```
Comment (mmusgrov) [inline blacktie/tx/src/main/cpp/HttpTxManager.cxx]: Well spotted.

I think we should also be removing the branch after a successful commit or abort. Do you agree?

File: blacktie/tx/src/test/cpp/TestTransactions.cxx (hunk 2 above)
Comment (zhfeng) [inline blacktie/tx/src/test/cpp/TestTransactions.cxx]: Thanks Mike,

the rcCnt2 > 0 means that the recovery_cb2 () has been called but it still needs to wait for removing the transaction logs. So I think the if condition should be
``` 
if ((nrecs = count_log_records()) < nrecs1)
    break;
```

File: blacktie/tx/src/main/cpp/HttpTxManager.cxx (hunk 3 above)
Comment (zhfeng) [inline blacktie/tx/src/main/cpp/HttpTxManager.cxx]: I agree and will take a look at the codes

File: blacktie/tx/src/main/cpp/HttpTxManager.cxx [hunk 4]
```diff
@@ -379,6 +379,7 @@ bool HttpTxManager::handle_request(
 					LOG4CXX_DEBUG(httptxlogger, "Return 200 OK with body: " <<
 						status << " XA status: " << res);
 					code = 200;
+					_branches.erase(branch->get_name());
```
Comment (zhfeng) [inline blacktie/tx/src/main/cpp/HttpTxManager.cxx]: I add  to remove the branch

Comment (mmusgrov) [inline blacktie/tx/src/main/cpp/HttpTxManager.cxx]: OK I think that should be enough.

File: blacktie/tx/src/main/cpp/HttpTxManager.cxx (hunk 3 above)
Comment (mmusgrov) [inline blacktie/tx/src/main/cpp/HttpTxManager.cxx]: This change is wrong (see Amos' comment on the associated JIRA JBTM-2363)

File: blacktie/tx/src/main/cpp/HttpTxManager.cxx (hunk 4 above)
Comment (mmusgrov) [inline blacktie/tx/src/main/cpp/HttpTxManager.cxx]: Similarly here too - this change is wrong (see Amos' comment on the associated JIRA JBTM-2363)

File: blacktie/tx/src/main/cpp/HttpTxManager.cxx (hunk 3 above)
Comment (zhfeng) [inline blacktie/tx/src/main/cpp/HttpTxManager.cxx]: I think this change looks good

File: blacktie/tx/src/main/cpp/HttpTxManager.cxx (hunk 4 above)
Comment (zhfeng) [inline blacktie/tx/src/main/cpp/HttpTxManager.cxx]: this change is wrong as the branch looks like it has been delete after calling the do_commit or do_rollback. I have a fix in https://github.com/jbosstm/narayana/pull/828

**Pull request merged**